rpisense-fb.c

```diff
@@ -21,247 +21,200 @@
-			   0x03, 0x03, 0x03, 0x04, 0x04, 0x05, 0x05, 0x06,
-			   0x06, 0x07, 0x07, 0x08, 0x08, 0x09, 0x0A, 0x0A,};
+static int rpisense_fb_probe(struct platform_device *pdev)
+{
+	int ret;
 
-static u8 gamma_user[32];
+	struct rpisense *rpisense = dev_get_drvdata(&pdev->dev);
+	struct rpisense_fb *rpisense_fb = &rpisense->framebuffer;
 
-static u32 pseudo_palette[16];
+	rpisense_fb->gamma = devm_kmalloc(&pdev->dev, 32, GFP_KERNEL);
+	if(!rpisense_fb->gamma)
+		return -ENOMEM;
+	memcpy(rpisense_fb->gamma, gamma_presets[lowlight], 32);
 
-static struct rpisense_fb_param rpisense_fb_param = {
-	.vmem = NULL,
-	.vmemsize = 128,
-	.gamma = gamma_default,
-};
+	rpisense_fb->vmem = (void *)devm_get_free_pages(&pdev->dev, GFP_KERNEL, 0);
+	if (!rpisense_fb->vmem)
+		return -ENOMEM;
+	memset(rpisense_fb->vmem, 0, PAGE_SIZE);
 
-static struct fb_deferred_io rpisense_fb_defio;
+	mutex_init(&rpisense_fb->rw_mtx);
 
-static struct fb_fix_screeninfo rpisense_fb_fix = {
-	.id =		"RPi-Sense FB",
-	.type =		FB_TYPE_PACKED_PIXELS,
-	.visual =	FB_VISUAL_TRUECOLOR,
-	.xpanstep =	0,
-	.ypanstep =	0,
-	.ywrapstep =	0,
-	.accel =	FB_ACCEL_NONE,
-	.line_length =	16,
-};
+	rpisense_fb->mdev = (struct miscdevice) {
+		.minor	= MISC_DYNAMIC_MINOR,
+		.name	= "sense-hat",
+		.mode	= 0666,
+		.fops	= &rpisense_fb_fops,
+	};
 
-static struct fb_var_screeninfo rpisense_fb_var = {
-	.xres		= 8,
-	.yres		= 8,
-	.xres_virtual	= 8,
-	.yres_virtual	= 8,
-	.bits_per_pixel = 16,
-	.red		= {11, 5, 0},
-	.green		= {5, 6, 0},
-	.blue		= {0, 5, 0},
-};
+	ret = misc_register(&rpisense_fb->mdev);
+	if (ret < 0) {
+		dev_err(&pdev->dev, "Could not register framebuffer.\n");
+		return ret;
+	}
 
-static ssize_t rpisense_fb_write(struct fb_info *info,
-				 const char __user *buf, size_t count,
-				 loff_t *ppos)
-{
-	ssize_t res = fb_sys_write(info, buf, count, ppos);
+	dev_info(&pdev->dev, "framebuffer registered with minor number %i", rpisense_fb->mdev.minor);
 
-	schedule_delayed_work(&info->deferred_work, rpisense_fb_defio.delay);
-	return res;
+	rpisense_update_framebuffer(rpisense);
+	return 0;
 }
 
-static void rpisense_fb_fillrect(struct fb_info *info,
-				 const struct fb_fillrect *rect)
+static int rpisense_fb_remove(struct platform_device *pdev)
 {
-	sys_fillrect(info, rect);
-	schedule_delayed_work(&info->deferred_work, rpisense_fb_defio.delay);
+	struct rpisense *rpisense = dev_get_drvdata(&pdev->dev);
+	struct rpisense_fb *rpisense_fb = &rpisense->framebuffer;
+	misc_deregister(&rpisense_fb->mdev);
+	return 0;
 }
 
-static void rpisense_fb_copyarea(struct fb_info *info,
-				 const struct fb_copyarea *area)
+static loff_t rpisense_fb_llseek(struct file *filp, loff_t pos, int whence)
 {
-	sys_copyarea(info, area);
-	schedule_delayed_work(&info->deferred_work, rpisense_fb_defio.delay);
+	loff_t base;
+	switch(whence)
+	{
+	case SEEK_SET:
+		base = 0;
+		break;
+	case SEEK_CUR:
+		base = filp->f_pos;
+		break;
+	case SEEK_END:
+		base = 128;
```

Comment:
I'd #define 128 since it appears a lot

Reply:
when I switch to more strong types for the code I can probably use sizeof